### PR TITLE
Spmv

### PR DESCRIPTION
**Pull request by NMerz**

### CODE REVIEW

**Comment by pandeyshubham25** [comment]
have completed the implementation for COO SPMV, also added a piece of code to verify base ranks with other and compute RMSE

File: src/prank_trivial_parallel.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                for(vector <int> :: iterator it = pages[idx].in_ids.begin(); it != pages[idx].in_ids.end(); ++it)
+                    pages[idx].score_new += pages[*it].score / pages[*it].outdegree;
+            }
+
+            // apply pageRank equation
+            pages[idx].score_new = alpha * pages[idx].score_new + (1.0 - alpha) / num_pages;
+        }
+
+        // update broadcast score
+        broadcastScore = alpha * broadcastScore / num_pages;
```
Comment (NMerz) [inline src/prank_trivial_parallel.cpp]: Is there an equivalent for this in the spmv implementation? If not, will that cause us problems?

File: src/prank_parallel_spmv.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    }
+    cout<<"Algorithm terminated in "<<(double)(clock() - tStart)/CLOCKS_PER_SEC<<"s"<<endl;
+
+    //PHASE 4: Write the result
+    cout<<"Writing result..."<<endl;
+    string result_path = input_file + string("_parallel_ranks") ;
+    FILE *fptr = fopen(result_path.c_str(), "w");
+    for(int i=0; i<cols; i++){
+        fprintf(fptr, "%d %.10f\n", i+1, C0[i]);
```
Comment (NMerz) [inline src/prank_parallel_spmv.cpp]: trivial parallel and serial are 0 indexed.

Comment (pandeyshubham25) [reply src/prank_parallel_spmv.cpp]: my bad, will fix this. perhaps this will make my RMSE a little better !

**Comment by NMerz** [comment]
Added CSR support. Tested vs serial on a couple tiny graphs:
<details>
  <summary>Example1 (click to expand)</summary>

```
0 0
0 1
1 1
1 2
2 0
3 3
```

</details>
<details>
  <summary>Example2 (click to expand)</summary>

```
0 0
0 1
0 3
1 1
1 2
2 0
2 3
3 3
```

</details>

File: src/prank_trivial_parallel.cpp (hunk 1 above)
Comment (pandeyshubham25) [reply src/prank_trivial_parallel.cpp]: I am not sure what precisely is the equivalent for this in spmv, I simply follwed the matrix multiplication rule mentioned in that paper. So we should be fine with that.

Comment (NMerz) [reply src/prank_trivial_parallel.cpp]: I think this is significant and will cause problems. See the sink section [here](http://cs.brown.edu/courses/cs016/static/files/assignments/projects/GraphHelpSession.pdf)

Also a small example:
```
(base) nmerz@Nathans-MacBook-Pro pagerank % cat csr_test3
0 0
0 1
1 1
1 2
1 3
2 0
```
```
(base) nmerz@Nathans-MacBook-Pro pagerank % diff <(sort csr_test3_csr_parallel_ranks) <(sort csr_test3_serial_ranks)
1,4c1,4
< 0 0.2777376473
< 1 0.2389857173
< 2 0.1903816313
< 3 0.1903816313
---
> 0 0.3094615782
> 1 0.2662833242
> 2 0.2121275480
> 3 0.2121275480
```

Comment (pandeyshubham25) [reply src/prank_trivial_parallel.cpp]: Got it, lemme check this one and get back

**Comment by pandeyshubham25** [comment]
Made a couple of minor changes -
1) Need your help with adding a few more descriptions in Readme.md, have added "@Nathan" in the file where you can fill in more content.
2) Got rid of prank_trivial_parallel file and copied the content to prank_parallel file instead
3) prank_serial used to take num_iters as first argument, changed to to only take the args relevant to it

**Pull request merged**